route_following_plugin/src/route_following_plugin.cpp

```diff
@@ -41,190 +41,210 @@ namespace route_following_plugin
+        
         discovery_pub_timer_ = pnh_->createTimer(
             ros::Duration(ros::Rate(10.0)),
             [this](const auto&) { plugin_discovery_pub_.publish(plugin_discovery_msg_); });
     }
+
     void RouteFollowingPlugin::run()
     {
         initialize();
         ros::CARMANodeHandle::spin();
     }
-    bool RouteFollowingPlugin::plan_maneuver_cb(cav_srvs::PlanManeuversRequest &req, cav_srvs::PlanManeuversResponse &resp)
-    {        
-        lanelet::BasicPoint2d current_loc(pose_msg_.pose.position.x, pose_msg_.pose.position.y);
-        auto current_lanelets = lanelet::geometry::findNearest(wm_->getMap()->laneletLayer, current_loc, 10);       
-        if(current_lanelets.size() == 0)
+
+    void RouteFollowingPlugin::pose_cb(const geometry_msgs::PoseStampedConstPtr& msg)
+    {
+        pose_msg_ = geometry_msgs::PoseStamped(*msg.get());
+    }
+    void RouteFollowingPlugin::twist_cb(const geometry_msgs::TwistStampedConstPtr& msg)
+    {
+        current_speed_ = msg->twist.linear.x;
+    }
+
+    std::vector<cav_msgs::Maneuver> RouteFollowingPlugin::route_cb(lanelet::routing::LaneletPath route_shortest_path, lanelet::BasicPoint2d& current_loc){
+        std::vector <cav_msgs::Maneuver> maneuvers;
+        //This function calculates the maneuver plan every time the route is set
+
+        //Go through entire route - identify lane changes and fill in the spaces with lane following
+        auto nearest_lanelets = lanelet::geometry::findNearest(wm_->getMap()->laneletLayer, current_loc, 10); 
+        if(nearest_lanelets.size() == 0)
         {
             ROS_WARN_STREAM("Cannot find any lanelet in map!");
-            return true;
         }
-        
-        auto shortest_path = wm_->getRoute()->shortestPath();
 
-        lanelet::ConstLanelet current_lanelet;
-        int last_lanelet_index = -1;
-        for (auto llt : current_lanelets)
+        //auto shortest_path = wm_->getRoute()->shortestPath();
+
+        //Check if current position is on shortest path, if not on shortest path return empty maneuver  
+        int current_lanelet_index = -1;
+        for (auto llt : nearest_lanelets)
         {
             if (boost::geometry::within(current_loc, llt.second.polygon2d()))
             {
-                int potential_index = findLaneletIndexFromPath(llt.second.id(), shortest_path);
+                int potential_index = findLaneletIndexFromPath(llt.second.id(), route_shortest_path);
                 if (potential_index != -1)
                 {
-                    last_lanelet_index = potential_index;
-                    current_lanelet = shortest_path[last_lanelet_index];
+                    current_lanelet_index = potential_index;
                     break;
                 }
             }
         }
-        if(last_lanelet_index == -1)
+        if(current_lanelet_index == -1)
         {
             ROS_ERROR_STREAM("Current position is not on the shortest path! Returning an empty maneuver");
```

Comment:
I think the design needs to be reconsidered for situations where vehicle is off the shortest path.  In the meantime, the quick solution is to change this error statement to a warning and remove the second sentence (since there is no return statement here), and just let the method continue building a maneuvers list as it does.  This will allow steering control to bring things back to the shortest path eventually.